File: native/ios/Sources/PasswordRecovery.swift
import Combine
import Foundation

@MainActor
protocol PasswordRecoveryService {
    func requestReset(username: String) async throws -> Bool
    func confirmReset(username: String, password: String, code: String) async throws
}

enum RecoveryStep: Hashable { case closed, request, confirm, complete }
struct RecoveryState: Equatable {
    var step: RecoveryStep = .closed
    var busy = false
    var failed = false
}

@MainActor
final class PasswordRecovery: ObservableObject {
    @Published private(set) var state = RecoveryState()
    private let service: PasswordRecoveryService?
    private var username: String?
    private var operation: Task<Void, Never>?
    private var generation = 0

    init(service: PasswordRecoveryService?) { self.service = service }
    func close() {
        generation += 1
        operation?.cancel()
        operation = nil
        username = nil
        state = RecoveryState()
    }
    func open() {
        close()
        if service != nil { state = RecoveryState(step: .request) }
    }
    func request(_ value: String) {
        let trimmed = value.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let service, !trimmed.isEmpty, state.step == .request, !state.busy else { return }
        username = trimmed
        run { [weak self] in
            let complete = try await service.requestReset(username: trimmed)
            try Task.checkCancellation()
            if complete { self?.username = nil }
            self?.state = RecoveryState(step: complete ? .complete : .confirm)
        }
    }
    func confirm(password: String, code: String) {
        let trimmed = code.trimmingCharacters(in: .whitespacesAndNewlines)
        guard let service, let username, !password.isEmpty, !trimmed.isEmpty,
              state.step == .confirm, !state.busy else { return }
        run { [weak self] in
            try await service.confirmReset(username: username, password: password, code: trimmed)
            try Task.checkCancellation()
            self?.username = nil
            self?.state = RecoveryState(step: .complete)
        }
    }
    private func run(_ block: @escaping @MainActor () async throws -> Void) {
        generation += 1
        let current = generation
        state.busy = true
        state.failed = false
        operation = Task { [weak self] in
            do { try await block() }
            catch is CancellationError { return }
            catch {
                guard let self, current == generation, !Task.isCancelled else { return }
                state.failed = true
            }
            if let self, current == generation { state.busy = false }
        }
    }
}

File: native/android/app/src/main/java/com/mediconnect/nativeapp/PasswordRecovery.kt
package com.mediconnect.nativeapp

import kotlinx.coroutines.CancellationException
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Job
import kotlinx.coroutines.currentCoroutineContext
import kotlinx.coroutines.ensureActive
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch

interface PasswordRecoveryService {
    /** True means the SDK reports completion; false means confirmation is required. */
    suspend fun requestReset(username: String): Boolean
    suspend fun confirmReset(username: String, password: String, code: String)
}

enum class RecoveryStep { CLOSED, REQUEST, CONFIRM, COMPLETE }
data class RecoveryState(val step: RecoveryStep = RecoveryStep.CLOSED, val busy: Boolean = false, val failed: Boolean = false)

class PasswordRecovery(private val service: PasswordRecoveryService?, private val scope: CoroutineScope) {
    private val mutable = MutableStateFlow(RecoveryState())
    val state = mutable.asStateFlow()
    private var username: String? = null
    private var operation: Job? = null
    private var generation = 0L

    fun close() {
        generation++
        operation?.cancel()
        operation = null
        username = null
        mutable.value = RecoveryState()
    }
    fun open() {
        close()
        if (service != null) mutable.value = RecoveryState(RecoveryStep.REQUEST)
    }
    fun request(value: String) {
        val service = service ?: return
        val trimmed = value.trim()
        if (trimmed.isEmpty() || mutable.value.step != RecoveryStep.REQUEST || mutable.value.busy) return
        username = trimmed
        run {
            val complete = service.requestReset(trimmed)
            currentCoroutineContext().ensureActive()
            if (complete) username = null
            mutable.value = RecoveryState(if (complete) RecoveryStep.COMPLETE else RecoveryStep.CONFIRM)
        }
    }
    fun confirm(password: String, code: String) {
        val service = service ?: return
        val username = username ?: return
        if (password.isEmpty() || code.isBlank() || mutable.value.step != RecoveryStep.CONFIRM || mutable.value.busy) return
        run {
            service.confirmReset(username, password, code.trim())
            currentCoroutineContext().ensureActive()
            this.username = null
            mutable.value = RecoveryState(RecoveryStep.COMPLETE)
        }
    }
    private fun run(block: suspend () -> Unit) {
        val current = ++generation
        mutable.value = mutable.value.copy(busy = true, failed = false)
        operation = scope.launch {
            try { block() }
            catch (cancelled: CancellationException) { throw cancelled }
            catch (_: Exception) {
                if (current == generation) mutable.value = mutable.value.copy(failed = true)
            } finally {
                if (current == generation) mutable.value = mutable.value.copy(busy = false)
            }
        }
    }
}
